auth_mellon_diagnostics.c

```diff
@@ -0,0 +1,1023 @@
+                        "varname=\"%s\" flags=%s str=\"%s\" directive=\"%s\"",
+                        cond->varname, am_diag_cond_flag_str(r, cond->flags),
+                        cond->str, cond->directive);
+}
+
+const char *
+am_diag_cache_key_type_str(am_cache_key_t key_type)
+{
+    switch(key_type) {
+    case AM_CACHE_SESSION: return "session";
+    case AM_CACHE_NAMEID : return "name id";
+    default:               return "unknown";
+    }
+}
+
+const char *
+am_diag_lasso_http_method_str(LassoHttpMethod http_method)
+{
+    switch(http_method) {
+    case LASSO_HTTP_METHOD_NONE:          return "LASSO_HTTP_METHOD_NONE";
+    case LASSO_HTTP_METHOD_ANY:           return "LASSO_HTTP_METHOD_ANY";
+    case LASSO_HTTP_METHOD_IDP_INITIATED: return "LASSO_HTTP_METHOD_IDP_INITIATED";
+    case LASSO_HTTP_METHOD_GET:           return "LASSO_HTTP_METHOD_GET";
+    case LASSO_HTTP_METHOD_POST:          return "LASSO_HTTP_METHOD_POST";
+    case LASSO_HTTP_METHOD_REDIRECT:      return "LASSO_HTTP_METHOD_REDIRECT";
+    case LASSO_HTTP_METHOD_SOAP:          return "LASSO_HTTP_METHOD_SOAP";
+    case LASSO_HTTP_METHOD_ARTIFACT_GET:  return "LASSO_HTTP_METHOD_ARTIFACT_GET";
+    case LASSO_HTTP_METHOD_ARTIFACT_POST: return "LASSO_HTTP_METHOD_ARTIFACT_POST";
+    case LASSO_HTTP_METHOD_PAOS:          return "LASSO_HTTP_METHOD_PAOS";
+    default:                              return "unknown";
+    }
+}
+
+void
+am_diag_printf(request_rec *r, const char *fmt, ...)
+{
+    va_list ap;
+    am_diag_cfg_rec *diag_cfg = am_get_diag_cfg(r->server);
+    am_req_cfg_rec *req_cfg = am_get_req_cfg(r);
+    char *buf;
+    apr_size_t buf_len;
+
+    if (!AM_DIAG_ENABLED(diag_cfg)) return;
+    if (!am_diag_initilaize_req(r, diag_cfg, req_cfg)) return;
+
+
+    va_start(ap, fmt);
+    buf = apr_pvsprintf(r->pool, fmt, ap);
+    va_end(ap);
+    buf_len = strlen(buf);
+    if (buf_len > 0) {
+        apr_file_write_full(diag_cfg->fd, buf, buf_len, NULL);
+    }
+    apr_file_flush(diag_cfg->fd);
+}
+
+void
+am_diag_rerror(const char *file, int line, int module_index,
+               int level, apr_status_t status,
+               request_rec *r, const char *fmt, ...)
+{
+    va_list ap;
+    am_diag_cfg_rec *diag_cfg = am_get_diag_cfg(r->server);
+    am_req_cfg_rec *req_cfg = am_get_req_cfg(r);
+    char *buf;
+    apr_size_t buf_len;
+
+    if (!AM_DIAG_ENABLED(diag_cfg)) return;
+    if (!am_diag_initilaize_req(r, diag_cfg, req_cfg)) return;
+
+
+    buf = apr_psprintf(r->pool, "[%s %s:%d] ",
+                       am_diag_httpd_error_level_str(r, level), file, line);
+    buf_len = strlen(buf);
+    if (buf_len > 0) {
+        apr_file_write_full(diag_cfg->fd, buf, buf_len, NULL);
+    }
+
+    va_start(ap, fmt);
+    buf = apr_pvsprintf(r->pool, fmt, ap);
+    va_end(ap);
+    buf_len = strlen(buf);
+    if (buf_len > 0) {
+        apr_file_write_full(diag_cfg->fd, buf, buf_len, NULL);
+    }
+    apr_file_flush(diag_cfg->fd);
```

Comment:
When looking at the diagnostics output, I noticed that these lines don't get a trailing newline, which causes a bit confusing log output. Since the convention for `ap_log_rerror()` is to not include a newline, I think this function will have to supply it.

Reply:
Good observation, yes you're correct. Fixed.